### DISCUSSION TITLE
Content with no length not captured in stream file and not saved via save_streamed_data.py

### DISCUSSION THREAD

**Discussion by EndUser509** [mitmproxy/mitmproxy#5460] [General]
#### Problem Description
It seems that documents that have no length specified do not have their body captured in stream files. They are also not streamed and as such cannot be saved with the `save_streamed_data.py` addon. 

If I replace the this check in `check_body_size()` in `mitmproxy/proxy/layers/http/__init_.py`

```
        if expected_size is None or expected_size <= 0:
            return False
```

with this (copy-paste from 'Step 3' further below in the code)

```
        if expected_size is None or expected_size <= 0:
            if request:
                self.flow.request.stream = True
                if self.request_body_buf:
                    # clear buffer and then fake a DataReceived event with everything we had in the buffer so far.
                    body_buf = self.request_body_buf
                    self.request_body_buf = b""
                    yield from self.start_request_stream()
                    yield from self.handle_event(RequestData(self.stream_id, body_buf))
            if response:
                assert self.flow.response
                self.flow.response.stream = True
                if self.response_body_buf:
                    body_buf = self.response_body_buf
                    self.response_body_buf = b""
                    yield from self.start_response_stream()
                    yield from self.handle_event(ResponseData(self.stream_id, body_buf))
            return False
```
then this forces streaming of such content and `save_streamed_data.py` does save the body.

What is the idea of the current behavior if it is intentional? If the body is not in the stream file then it is probably also not available in the `response()` hook which would be the only reason not to treat uknown size documents as streamed by default. Or would it make more sense to have an `--stream_unknown_size` option?

**Comment by mhils** [comment]
Moving this to discussions as per https://github.com/mitmproxy/mitmproxy/blob/main/examples/contrib/README.md#maintenance :)

**Comment by EndUser509** [comment]
Having this as a discussion may be better. I posted it as a bug because I see no way how to record unknown size bodies one way or another  without my ugly patch. :)

**Reply by mhils** [reply]
I think the problem is mostly the save_streamed_data addon itself, which incorrectly expects `response.stream` to be set in the responseheaders hook, even though it may only be set later in the case of chunked transfers (where we only later discover that body_size > stream_large_bodies[^1]). The addon could detect itself that the body size is indeterminate, and then always force streaming.

[^1]: This is why check_body_size is called repeatedly:
https://github.com/mitmproxy/mitmproxy/blob/821859cd4fccf754f6916a67b6dcb59bb4da2811/mitmproxy/proxy/layers/http/__init__.py#L423-L428 https://github.com/mitmproxy/mitmproxy/blob/821859cd4fccf754f6916a67b6dcb59bb4da2811/mitmproxy/proxy/layers/http/__init__.py#L523-L531 https://github.com/mitmproxy/mitmproxy/blob/821859cd4fccf754f6916a67b6dcb59bb4da2811/mitmproxy/proxy/layers/http/__init__.py#L572-L591

**Comment by EndUser509** [comment]
Thanks a lot for the pointer. I guess I could have debugged that further myself but I didn't realize that the check was done repeatedly. The case that I tested did in fact have a size that was larger than `stream_large_bodies`. 
I'll try to get the addon to deal with the late setting of the stream and post a PR.

**Comment by EndUser509** [comment]
It's still a little unfortunate that any addon that processes streams (such as `stream_modify.py` too) only has the option to force streaming for all unknown size content. 
I think the weakness here is the dual-use of the `flow.{request|response}.stream` member as both boolean and function object. It would be a lot easier to deal with this case if there was an additional member like `stream_processor` holding the function object that could be set in `{request|response}headers` _just in case it will be needed_. This could even be a set or dict where multiple addons can register their processor but maybe that is an overkill until somebody actually wants to use `stream_modify.py` _and_ `save_streamed_data.py` at the same time.
Would a PR that implements the above be welcome?

**Reply by mhils** [reply]
> It's still a little unfortunate that any addon that processes streams (such as `stream_modify.py` too) only has the option to force streaming for all unknown size content. I think the weakness here is the dual-use of the `flow.{request|response}.stream` member as both boolean and function object. It would be a lot easier to deal with this case if there was an additional member like `stream_processor` holding the function object that could be set in `{request|response}headers` _just in case it will be needed_. [...] Would a PR that implements the above be welcome?

I think so, yes. The `.stream` design predates the current behavior (there was a time where unknown sizes were always entirely streamed). 
Fundamentally speaking, it might be more useful to add an option for streaming requests/responses but at the same time retaining their bodies.

**Reply by EndUser509** [reply]
> I think so, yes. The `.stream` design predates the current behavior (there was a time where unknown sizes were always entirely streamed). 

One last question before I take a go at this: Having a proper API hook, maybe similar to `tcp_message`, is ultimately not the way to go? I have no opinion here, but it would be more consistent.

> Fundamentally speaking, it might be more useful to add an option for streaming requests/responses but at the same time retaining their bodies.

If the streamed data is really large and you try to keep it in memory to just dump it in the stream file will cause its own problems, if that is what you meant. Running out of memory and unresponsiveness when flushing a multi-GB download to the stream file for instance. 
Personally I use a script to extract all body data from stream files and mix it with the files written by the `save_streamed_data.py` plugin. I just want all requests and responses as individual files for my purpose. It would be relatively straight-forward to write a tool that does the opposite, merging the saved files into the stream file.
